limit rate of webxdc updates

r10s opened the pull request:
this pr is based based on the ratelimited introduced by https://github.com/deltachat/deltachat-core-rust/pull/3402:

- instead of sending the status-update-message directly, we now just interrupt smtp; the real status-update-message is then created in the smtp-loop (similar to sync-messages and mdn) and interrupting smtp.

- ratelimit is then handled in the same, already implemented way, resulting in smtp being interrupted again if there is another chance that something can be sent.

- as a nice side-effect, also sync-messages are delayed as well when ratelimit happens.

closes #3311

r10s commented on src/webxdc.rs:
```diff
@@ -367,10 +367,54 @@ impl Context {
+                async_std::future::timeout(duration_until_can_send, async {
+                    // TODO: this is not correct timeout() is executed directly (copy+paste error from scheduler :)
+                    // what we really want is that the following lines are exexcuted after the given duration
+                    if let Err(err) = self.flush_status_updates(instance).await {
```
@link2xt can you say what a good rust approach would be to have `flush_status_updates()` being executed after `duration_until_can_send`?

link2xt replied:
You can use `async_std::task::spawn` to create a new task, and `async_std::task::sleep` inside of it before calling `flush_status_updates`.

Better approach IMO is to modify long running SMTP loop so it flushes all updates right before sending messages, instead of spawning new tasks that nobody keeps track of.

r10s replied:
i finally went for the "modify long running SMTP loop" approach as there are several tiny issues with the task::spawn this seems to be even easier overall.

r10s commented on src/webxdc.rs:
```diff
@@ -340,7 +341,7 @@ impl Context {
         update_str: &str,
         descr: &str,
-    ) -> Result<Option<MsgId>> {
+    ) -> Result<()> {
```
thanks, i fixed that.

r10s commented on src/webxdc.rs:
```diff
@@ -365,35 +364,98 @@ impl Context {
+    /// allow to send them. Returns false if there are no more status updates to send.
+    pub(crate) async fn flush_status_updates(&self) -> Result<bool> {
+        loop {
+            if !self.ratelimit.read().await.can_send() {
```
bypassing `smtp` queue would be quite some work as we cannot easily reuse the rendering, ephemeral things and so on. therefore, i went the easy way, to create just a normal message.

but yes, checking ratelimit in the loop does not make sense then. but if we check outside the loop it is still an improvement as at least no more messages are sent if the ratelimit is triggered. in practise, most times there won't be anything to look anyway as there is only one message.

i do a commit that removes the ratelimit outside the loop, maybe we can see in a subsequent pr how hard bypassing smtp queue really is.

r10s replied:
i added a comment to the source code.

r10s commented on src/smtp.rs:
```diff
@@ -529,9 +534,12 @@ pub(crate) async fn send_smtp_messages(context: &Context, connection: &mut Smtp)
-    let ratelimited = send_mdns(context, connection)
-        .await
-        .context("failed to send MDNs")?;
+    if !ratelimited {
```
but we still need the return value to be `true` as otherwise the following may happen:
- webxdc-updates are not sent out due to ratelimit
- sending the other messages takes some time
- mdns are send out as ratelimit expired in between
in this case, the caller will not wait for `until_can_send` and the webxdc updates are not send in time. also, this would send mdn with a sometimes higher priority than webxdc updates.

r10s replied:
i added a comment to the source code.

r10s commented on src/smtp.rs:
```diff
@@ -506,7 +506,12 @@ async fn send_mdns(context: &Context, connection: &mut Smtp) -> Result<bool> {
 pub(crate) async fn send_smtp_messages(context: &Context, connection: &mut Smtp) -> Result<bool> {
-    context.send_sync_msg().await?; // Add sync message to the end of the queue if needed.
+    // If needed, add status updates and sync messages to end of sending queue
+    let mut ratelimited = context.flush_status_updates().await?;
```
see comment above, i think it makes sense not to try out sending mdn if ratelimite happens on sending webxdc-updates/sync. i think, it is better to retry sending webxdc-updates/sync as soon as possible.

r10s replied:
> No need for this ratelimited variable, you can just do [...]

did so and added a comment.

r10s commented on src/webxdc.rs:
```diff
@@ -365,35 +364,97 @@ impl Context {
+                    }
+                    self.sql
+                        .execute(
+                            "UPDATE smtp_status_updates SET last_serial=?, descr=? WHERE msg_id=?",
```
this is a clever sql statement. however, wondering if the mutex is still needed because of SELECT/DELETE needs to be executed together. or we would need to capsule that into an `IMMEDIATE TRANSACTION`

r10s replied:
there seems to be sth. as `DELETE ... RETURNING`, https://www.sqlite.org/lang_returning.html, that could do the job. it is pretty new in sqlite, available since 3.35.0 (2021-03-12), however, iirc, we ship our own sqlite, so this should not be an issue.

r10s replied:
i added commits to use `DELETE ... RETURNING` and removing the mutex.

r10s commented:
@link2xt thanks a lot for the detailed review, in incorporated lots of your ideas. especially the sql-part is much easier now :)

for the ratelimit in `send_smtp_messages()`: i played around with the idea of letting webxdc-updates- and [sync-messages](https://github.com/deltachat/deltachat-core-rust/pull/2669) bypass  the normal sending queue.  however, that seems to be quite some work and would probably double some code code and/or needs lots of refactoring.

i think the risk in adding new issues there is much larger than the effect - most times, there is only one status-update/sync anyway.

so, to me, it seems that the current approach in `send_smtp_messages()` is "good enough", however, i added some comments and removed the ratelimit check out of loops.
